src/EventStore.Core/DataStructures/ProbabilisticFilter/FileStreamPersistence.cs

```diff
@@ -0,0 +1,292 @@
+					$"the actual file size ({bulkFileStream.Length:N0}) of file {_path}");
+
+			var bytesToRead = DataAccessor.FileSize;
+			var bytesRead = 0L;
+			while (bytesToRead > 0) {
+				var bytesToReadInBlock = bytesToRead > int.MaxValue
+					? int.MaxValue
+					: (int)bytesToRead;
+
+				bulkFileStream.Read(new Span<byte>(DataAccessor.Pointer + bytesRead, bytesToReadInBlock));
+
+				bytesRead += bytesToReadInBlock;
+				bytesToRead -= bytesToReadInBlock;
+			}
+
+			var elapsed = sw.Elapsed;
+			var fileSizeMb = DataAccessor.FileSize / 1000 / 1000;
+			var megaBytesPerSecond = fileSizeMb / elapsed.TotalSeconds;
+			Log.Information(
+				"Read persisted bloom filter {path} into memory. Took {elapsed}. {megaBytesPerSecond:N2} MB/s",
+				_path,
+				elapsed,
+				megaBytesPerSecond);
+		}
+
+		private void OnPageDirty(long pageNumber) {
+			lock (_bitmapLock) {
+				ThrowIfDisposed();
+				var byteIndex = (int)(pageNumber / 8);
+				var bitIndex = pageNumber % 8;
+				ref var byteValue = ref _dirtyPageBitmap.AsSpan()[byteIndex];
+				byteValue = byteValue.SetBit(bitIndex);
+			}
+		}
+
+		public void Flush() {
+			using var fileStream = new FileStream(
+				_path,
+				FileMode.OpenOrCreate,
+				FileAccess.ReadWrite,
+				FileShare.ReadWrite,
+				bufferSize: DataAccessor.PageSize);
+
+			fileStream.SetLength(DataAccessor.FileSize);
+
+			Span<byte> localCacheLine = stackalloc byte[BloomFilterIntegrity.CacheLineSize];
+			localCacheLine.Clear();
+
+			var pageNumber = 0L;
+			var flushedPages = 0L;
+			var pauses = 0;
+
+			var activelyFlushing = Stopwatch.StartNew();
+
+			for (var remaining = _dirtyPageBitmap.AsSpan();
+				remaining.Length > 0;
+				remaining = remaining[BloomFilterIntegrity.CacheLineSize..]) {
+
+				lock (_bitmapLock) {
+					ThrowIfDisposed();
+					var cacheLine = remaining[..BloomFilterIntegrity.CacheLineSize];
+					cacheLine.CopyTo(localCacheLine);
+					cacheLine.Clear();
+				}
+
+				foreach (var @byte in localCacheLine) {
+					for (var bitOffset = 0; bitOffset < 8; bitOffset++) {
+						if (@byte.IsBitSet(bitOffset)) {
+							WritePage(pageNumber, fileStream);
+							flushedPages++;
+
+							if (flushedPages % FlushBatchSize == 0) {
+								fileStream.FlushToDisk();
+								activelyFlushing.Stop();
+								pauses++;
+								Thread.Sleep(FlushBatchDelay);
+								activelyFlushing.Start();
+							}
+						}
+
+						pageNumber++;
+						if (pageNumber == DataAccessor.NumPages)
+							goto Done;
+					}
+				}
+			}
+
+			Done:
+			fileStream.FlushToDisk();
+
+			activelyFlushing.Stop();
+
+			var flushedBytes = flushedPages * DataAccessor.PageSize;
+			var flushedMegaBytes = (float)flushedBytes / 1000 / 1000;
+			var activeFlushRateMBperS = flushedMegaBytes / activelyFlushing.Elapsed.TotalSeconds;
+
+			Log.Information(
+				"Flushed {pages:N0} pages out of {totalPages:N0}. {bytes:N0} bytes. " +
+				"Delay {delay} per batch. Total delay {totalDelay:N0}ms. " +
```

Comment:
```suggestion
				"Delay {delay} ms per batch. Total delay {totalDelay:N0} ms. " +
```

nit: add unit and space